src/lib_cjp2/jp2.cpp

```diff
@@ -0,0 +1,112 @@
+#include "jp2.h"
+
+#include <cstdint>
+#include <fstream>
+#include <stdexcept>
+#include <vector>
+
+#include "codestream.h"
+#include "helper.h"
+
+void Box::read(std::ifstream &reader) {
+    reader_ = &reader;
+
+    lbox_ = read_u32(reader);
+    if (lbox_ < 8 && lbox_ != 0 && lbox_ != 1) {
+        throw std::runtime_error("A box has incorrect length!");
+    }
+
+    tbox_ = read_u32(reader);
+
+    if (lbox_ == 1) {
+        xlbox_ = read_u64(reader);
+        if (xlbox_ < 16) {
+            throw std::runtime_error("A box has incorrect length!");
+        }
+
+        //	read_exact(box.dbox, reader, box.xlbox - 16);
+        //} else {
+        //	read_exact(box.dbox, reader, box.lbox == 0 ? reader.right : box.lbox - 8);
+    }
+
+    dbox_.assign((lbox_ == 1 ? xlbox_ - 16 : lbox_ - 8), 0);
+    dbox_start_ = reader.tellg();
+
+    reader.seekg(dbox_start_ + dbox_.size());
+}
+
+// template<typename T>
+// T Box::unbox() {
+//	return *reinterpret_cast<T*>(&dbox);
+// }
+
+template <typename T>
+T Box::unbox() {
+    reader->seekg(dbox_start_);
+    if (lbox == 1) {
+        read_exact(dbox_, *reader, xlbox);
+    } else {
+        read_exact(dbox_, *reader, lbox == 0 ? reader->right : lbox - 8);
+    }
+
+    return *reinterpret_cast<T *>(&buf);

```

Comment:
А здесь тогда кастовать непосредственно dbox
```suggestion
    return *reinterpret_cast<T *>(&dbox_);
```